### PR TITLE
Add JCache serialization tests

### PR DESCRIPTION
**Pull request by ryanesch**
We need to add version-to-version compatibility testing for some of the serialized classes that were affected for the distributed cache feature. The tests should fail if a breaking change is made between versions.

Tests included:

com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java
com.ibm.ws.security.token/src/com/ibm/ws/security/token/internal/AbstractTokenImpl.java
com.ibm.ws.security.token/src/com/ibm/ws/security/token/internal/SingleSignonTokenImpl.java
com.ibm.ws.security.token.ltpa/src/com/ibm/ws/security/token/ltpa/internal/LTPAToken2.java
com.ibm.ws.security.social/src/com/ibm/websphere/security/social/UserProfile.java
-includes com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/internal/ClaimsImpl.java

### CODE REVIEW

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
https://github.com/OpenLiberty/open-liberty/issues/20568

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import org.junit.Test;
+
+import com.ibm.ws.security.authentication.cache.CacheObject;
+import com.ibm.ws.security.authentication.principals.WSPrincipal;
+
+/**
+ * Ensure CacheObject and all serialized instances of CacheObject remain serializable.
+ * If CacheObject is changed, create a new CacheObject_x.ser file and write a new test.
+ * Each new version should be able to deserialze every previous version.
+ * This is necessary for the distributed cache.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]:
```suggestion
 * This is necessary to maintain version to version compatibility for the distributed authentication cache.
```

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            FileInputStream file = new FileInputStream(filename);
+            ObjectInputStream in = new ObjectInputStream(file);
+
+            CacheObject object = (CacheObject) in.readObject();
+
+            //Check serializationVersion via reflection
+            Field field = CacheObject.class.getDeclaredField("serializationVersion");
+            field.setAccessible(true);
+            Object value = field.get(object);
+            assertEquals((short) 1, value);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]: Global comment: add failure/expecation messages so we know what is wrong in the failed test.

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+            //Check lookup keys
+            List<Object> keys = object.getLookupKeys();
+            assertTrue("There should be two lookup keys.", keys.size() == 2);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]:
```suggestion
            assertEquals("There should be two lookup keys.", 2, keys.size());
```

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            assertEquals("[lookupkey1, lookupkey2]", keys.toString());
+
+            //Check principals
+            Set<Principal> principals = object.getSubject().getPrincipals();
+            assertTrue("There should be one principal.", principals.size() == 1);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]:
```suggestion
            assertEquals("There should be one principal.", 1, principals.size());
```

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            Principal principal = principals.iterator().next();
+            assertEquals("securityName1", principal.getName());
+
+            //Check private credentials
+            Hashtable<String, Object> privateCreds = (Hashtable<String, Object>) object.getSubject().getPrivateCredentials().iterator().next();
+            assertTrue("There should be one private credential.", privateCreds.size() == 1);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]:
```suggestion
            assertEquals("There should be one private credential.", 1, privateCreds.size());
```

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            assertEquals("{privatekey1=privatevalue1}", privateCreds.toString());
+
+            //Check public credentials
+            Hashtable<String, Object> publicCreds = (Hashtable<String, Object>) object.getSubject().getPublicCredentials().iterator().next();
+            assertTrue("There should be one public credential.", publicCreds.size() == 1);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]:
```suggestion
            assertEquals("There should be one public credential.", 1, publicCreds.size());
```

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            assertEquals("{publickey1=publicvalue1}", publicCreds.toString());
+
+        } catch (Exception ioe) {
+            fail(ioe.toString());
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]: I am not a fan of catching exceptions in unit tests. I typically think the exception being reported in the JUnit results is better as you get the entire stack trace etc.

File: dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    /**
+     * Method used to create and serialize the CacheObject (CacheObject_1.ser).
+     */
+    public void writeCacheObject() {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/test/com/ibm/ws/security/util/CacheObjectSerializationTest.java]: Do we call this from anywhere?

If not, could this not be a main method? If so update the comments to include steps on when / how to run it. Might also want to change the name of the file produced so it doesn't overwrite the existing file.

File: dev/com.ibm.ws.security.authentication/bnd.bnd [hunk 9]
```diff
@@ -38,7 +38,8 @@ Service-Component:\
 -buildpath: \
 	com.ibm.ws.logging;version=latest, \
 	com.ibm.ws.org.osgi.annotation.versioning;version=latest, \
-	com.ibm.ws.javaee.dd;version=latest
+	com.ibm.ws.javaee.dd;version=latest, \
+	com.ibm.ws.security.credentials;version=latest
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.authentication/bnd.bnd]: This shouldn't need to be added.

File: dev/com.ibm.ws.security.social/test/com/ibm/websphere/security/social/TestJwtToken.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * Test JwtToken to use in UserProfileSerializationTest.
+ */
+public class TestJwtToken implements JwtToken{
+
+	private static final long serialVersionUID = 1L;
+
+	@Override
+	public Claims getClaims() {
+		// TODO Auto-generated method stub
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.social/test/com/ibm/websphere/security/social/TestJwtToken.java]:
```suggestion
```

File: dev/com.ibm.ws.security.social/test/com/ibm/websphere/security/social/TestJwtToken.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	@Override
+	public String getHeader(String name) {
+		return "header1";
+	}
+
+	@Override
+	public String compact() {
+		// TODO Auto-generated method stub
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.social/test/com/ibm/websphere/security/social/TestJwtToken.java]:
```suggestion
```

File: dev/com.ibm.ws.security.social/test/com/ibm/websphere/security/social/UserProfileSerializationTest.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+import com.ibm.websphere.security.jwt.JwtException;
+import com.ibm.websphere.security.jwt.JwtToken;
+import com.ibm.websphere.security.jwt.KeyException;
+import com.ibm.ws.security.jwt.internal.ClaimsImpl;
+
+/**
+ * Ensure UserProfile and all serialized instances of UserProfile remain serializable.
+ * If UserProfile is changed, create a new UserProfile_x.ser file and write a new test.
+ * Each new version should be able to deserialze every previous version.
+ * This is necessary for the distributed cache.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.social/test/com/ibm/websphere/security/social/UserProfileSerializationTest.java]:
```suggestion
 * This is necessary to maintain version to version compatibility for the distributed authentication cache.
```

File: dev/com.ibm.ws.security.token.ltpa/test/com/ibm/ws/security/token/ltpa/internal/LTPAToken2SerializationTest.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+import test.UTLocationHelper;
+
+/**
+ * Ensure LTPAToken2 remains serializable. This is necessary because
+ * SingleSignonTokenImpl must remain serializable and SSOTokenImpl extends
+ * AbstractTokenImpl which includes Token which is an LTPAToken2.
+ * If LTPAToken2 is changed, create a new LTPAToken2_x.ser file and write a new test.
+ * Each new version should be able to deserialze every previous version.
+ * This is necessary for the distributed cache.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.token.ltpa/test/com/ibm/ws/security/token/ltpa/internal/LTPAToken2SerializationTest.java]:
```suggestion
 * This is necessary to maintain version to version compatibility for the distributed authentication cache.
```

File: dev/com.ibm.ws.security.token/test/com/ibm/ws/security/token/internal/SerializationTest.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+import com.ibm.ws.security.token.TokenService;
+
+/**
+ * Ensure SingleSignonTokenImpl/AbstractTokenImpl and all serialized instances remain serializable.
+ * If either TokenImpl is changed, create a new SingleSignonTokenImpl_x.ser file and write a new test.
+ * SingleSignonTokenImpl extends AbstractTokenImpl so we are testing both with SSOTI. AbstractTokenImpl
+ * includes a Token which is only implemented by LTPAToken2. The serialization test for LTPAToken2 is
+ * separate from this one (in com.ibm.ws.security.token.ltpa).
+ * Each new version should be able to deserialze every previous version.
+ * This is necessary for the distributed cache.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.token/test/com/ibm/ws/security/token/internal/SerializationTest.java]:
```suggestion
 * This is necessary to maintain version to version compatibility for the distributed authentication cache.
```

**Review by jvanhill — CHANGES_REQUESTED** [review]

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Review by jvanhill — APPROVED** [review]

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#run-libby-bot

**Pull request merged**